File: apps/nongyu-go-track-server/internal/sqlguard/sqlguard_test.go
package sqlguard

import (
	"errors"
	"strings"
	"testing"
)

func TestPrepare_AllowsSelectAndWith(t *testing.T) {
	t.Parallel()
	cases := []string{
		"SELECT metric_key, metric_value FROM daily_metrics LIMIT 10",
		"select count(*) as n from events where event_type = 'app_open'",
		"WITH x AS (SELECT stat_date FROM daily_metrics) SELECT * FROM x",
		"SELECT e.event_name FROM events e JOIN daily_dims d ON e.stat_date = d.stat_date",
	}
	for _, sql := range cases {
		got, err := Prepare(sql)
		if err != nil {
			t.Fatalf("%s: %v", sql, err)
		}
		if !strings.Contains(got.ExecSQL, "LIMIT 501") {
			t.Fatalf("expected wrap limit: %s", got.ExecSQL)
		}
	}
}

func TestPrepare_RejectsWritesAndForbiddenTables(t *testing.T) {
	t.Parallel()
	cases := []struct {
		sql  string
		want error
	}{
		{"DELETE FROM events", ErrNotSelect},
		{"SELECT 1; DROP TABLE events", ErrMultiStmt},
		{"SELECT * FROM events; SELECT 1", ErrMultiStmt},
		{"SELECT * FROM meta_jobs", ErrBadTable},
		{"SELECT * FROM events WHERE 1=1; ATTACH 'x' AS t", ErrMultiStmt},
		{"INSERT INTO events(event_id) VALUES ('x')", ErrNotSelect},
		{"SELECT * FROM events INTO dump", ErrForbidden},
		{"PRAGMA table_info(events)", ErrNotSelect},
		{"SELECT * FROM events --\n; DELETE FROM events", ErrMultiStmt},
	}
	for _, tc := range cases {
		_, err := Prepare(tc.sql)
		if err == nil {
			t.Fatalf("expected error for %q", tc.sql)
		}
		if !errors.Is(err, tc.want) {
			t.Fatalf("%q: got %v want %v", tc.sql, err, tc.want)
		}
	}
}

func TestPrepare_CommentDoesNotHideDelete(t *testing.T) {
	t.Parallel()
	_, err := Prepare("SELECT * FROM events /* */ ; DELETE FROM events")
	if !errors.Is(err, ErrMultiStmt) {
		t.Fatalf("got %v", err)
	}
}

func TestPrepare_EmptyAndTooLong(t *testing.T) {
	t.Parallel()
	if _, err := Prepare("   "); !errors.Is(err, ErrEmpty) {
		t.Fatalf("empty: %v", err)
	}
	long := "SELECT * FROM events WHERE x = '" + strings.Repeat("a", MaxSQLBytes) + "'"
	if _, err := Prepare(long); !errors.Is(err, ErrTooLong) {
		t.Fatalf("too long: %v", err)
	}
}

File: apps/nongyu-node-track-server/src/sqlguard/sqlguard.test.ts
import { describe, expect, it } from "vitest";
import {
  ErrBadTable,
  ErrEmpty,
  ErrForbidden,
  ErrMultiStmt,
  ErrNotSelect,
  ErrTooLong,
  MaxSQLBytes,
  isSqlGuardError,
  prepare,
} from "../sqlguard/index.js";

describe("sqlguard prepare", () => {
  it("allows select and with", () => {
    const cases = [
      "SELECT metric_key, metric_value FROM daily_metrics LIMIT 10",
      "select count(*) as n from events where event_type = 'app_open'",
      "WITH x AS (SELECT stat_date FROM daily_metrics) SELECT * FROM x",
      "SELECT e.event_name FROM events e JOIN daily_dims d ON e.stat_date = d.stat_date",
    ];
    for (const sql of cases) {
      const got = prepare(sql);
      expect(got.execSQL).toContain("LIMIT 501");
    }
  });

  it("rejects writes and forbidden tables", () => {
    const cases: Array<{ sql: string; want: Error }> = [
      { sql: "DELETE FROM events", want: ErrNotSelect },
      { sql: "SELECT 1; DROP TABLE events", want: ErrMultiStmt },
      { sql: "SELECT * FROM events; SELECT 1", want: ErrMultiStmt },
      { sql: "SELECT * FROM meta_jobs", want: ErrBadTable },
      { sql: "SELECT * FROM events; ATTACH 'x' AS t", want: ErrMultiStmt },
      { sql: "INSERT INTO events(event_id) VALUES ('x')", want: ErrNotSelect },
      { sql: "SELECT * FROM events INTO dump", want: ErrForbidden },
      { sql: "PRAGMA table_info(events)", want: ErrNotSelect },
      { sql: "SELECT * FROM events --\n; DELETE FROM events", want: ErrMultiStmt },
    ];
    for (const tc of cases) {
      expect(() => prepare(tc.sql)).toThrow();
      try {
        prepare(tc.sql);
      } catch (err) {
        expect(isSqlGuardError(err, tc.want)).toBe(true);
      }
    }
  });

  it("comment does not hide delete", () => {
    try {
      prepare("SELECT * FROM events /* */ ; DELETE FROM events");
      expect.fail("should throw");
    } catch (err) {
      expect(isSqlGuardError(err, ErrMultiStmt)).toBe(true);
    }
  });

  it("empty and too long", () => {
    try {
      prepare("   ");
      expect.fail("should throw");
    } catch (err) {
      expect(isSqlGuardError(err, ErrEmpty)).toBe(true);
    }
    const long = "SELECT * FROM events WHERE x = '" + "a".repeat(MaxSQLBytes) + "'";
    try {
      prepare(long);
      expect.fail("should throw");
    } catch (err) {
      expect(isSqlGuardError(err, ErrTooLong)).toBe(true);
    }
  });
});
